### PR TITLE
Introduce TypedStore<State,Action>!

### LINKED ISSUE #46: ReducerForAction not usable with current store
[There's a definition of a `ReducerForAction`](https://github.com/reduxkotlin/redux-kotlin/blob/57fae507daaf84982dc79db253abe7eee7c46544/lib/src/commonMain/kotlin/org/reduxkotlin/Definitions.kt#L13) which [cannot be used in order to create a store](https://github.com/reduxkotlin/redux-kotlin/blob/57fae507da/lib/src/commonMain/kotlin/org/reduxkotlin/CreateStore.kt#L31). A suggestion could be smth like this:

```kotlin
/**
 * See also https://github.com/reactjs/redux/blob/master/docs/Glossary.md#reducer
 */
typealias Reducer<State> = ReducerForActionType<State, Any>

/**
 * Reducer for a particular subclass of actions.  Useful for Sealed classes &
 * exhaustive when statements.  See [reducerForActionType].
 */
typealias ReducerForActionType<TState, TAction> = (state: TState, action: TAction) -> TState
```

and then make sure that the store can handle the `ReducerForActionType<TState, TAction>`

Personally I would change this the other way around and make the reducer be the one with the defined action:

```kotlin
/**
 * See also https://github.com/reactjs/redux/blob/master/docs/Glossary.md#reducer
 */
typealias Reducer<State, Action> = (state: State, action: Action) -> State

/**
 * ...
 */
typealias AnonymActionReducer<State> = Reducer<State, Any>
```

### LINKED ISSUE #92: createThreadSafeStore doesn't support ReducerForActionType
To create a store using `createThreadSafeStore`, you must provide:

```
typealias Reducer<State> = (state: State, action: Any) -> State
```
And not:

```
typealias ReducerForActionType<TState, TAction> = (state: TState, action: TAction) -> TState
```

Can we add support for the latter?

### PR DESCRIPTION
**Pull request by mpetuska**
Closes #92
Closes #46

### CODE REVIEW

**Comment by mpetuska** [comment]
Some additional tests are still missing but public API is final.

**Comment by patjackson52** [comment]
I'm not following why adding a type for action to Store.  What is the need this is fulfilling? To require a common subclass for all actions?  I'm in favor of not requiring a type for actions.   Most stores would use Any as action type and requiring a common subclass is just adding complexity & confusion imo.

The `ReducerForAction` type was used along with some other code as a way reduce type checking in reducers and make them more concise.  I think a full example of how it was used is lacking.   Not sure this is even a real issue.

**Comment by mpetuska** [comment]
To be clear, this feature set is entirely opt in. The benefits are typesafe dispatchers and reducers. The downsides are that some enchancers might become unavailable (e.g. thunk). In any case, traditional stores and reducers remain exactly the same. I'll bring back removed typealiases with deprecation in the next patch to this.

I personally missed typesafe store in some of my projects as well.

**Comment by mpetuska** [comment]
@patjackson52 PR is ready for review. Here are a few things I'd like you to double-check while reviewing:
- No existing API changes (i.e. old code will fully compile with new code)
- New `Typed` entities are entirely opt-in
- Two deprecated entities make sense and looks correct to work with Intellij auto-replace intents.

File: redux-kotlin/src/commonMain/kotlin/org/reduxkotlin/CreateStore.kt [hunk 1]
```diff
@@ -246,3 +246,26 @@ public fun <State> createStore(
...
+        enhancer,
+    )
+    return object : TypedStore<State, Action> {
+        override val getState: GetState<State> = store.getState
+        override var dispatch: TypedDispatcher<Action> = store.dispatch
+        override val subscribe: (StoreSubscriber) -> StoreSubscription = store.subscribe
+        override val replaceReducer: (TypedReducer<State, Action>) -> Unit = {
+            store.replaceReducer(typedReducer(it))
+        }
+    }
```
Comment (patjackson52) [inline redux-kotlin/src/commonMain/kotlin/org/reduxkotlin/CreateStore.kt:270]: Can this use `asTyped()`?
 = createStore(
        reducer = typedReducer(reducer),
        preloadedState,
        enhancer,
    ).asTyped()

**Review by patjackson52 — APPROVED** [review]
Lgtm, thanks for reworking and keeping original api unchanged.  Addition of TypedStore seems to be requested enough that should be included in the lib.
Thanks for putting this up!

File: redux-kotlin/src/commonMain/kotlin/org/reduxkotlin/CreateStore.kt (hunk 1 above)
Comment (mpetuska) [reply redux-kotlin/src/commonMain/kotlin/org/reduxkotlin/CreateStore.kt:270]: Good spot!

**Pull request merged**